File: Yoonsuh/PGS/등산코스 정하기.py
##### 다익스트라 알고리즘 템플릿 #####

import heapq

def solution(n, paths, gates, summits):
    hiking = [[] for _ in range(n+1)]
    dist = [float('inf')] * (n + 1)
    for path in paths:
        i, j, w = path
        hiking[i].append((j, w))
        hiking[j].append((i, w))
    
    pq = []
    for gate in gates:
        heapq.heappush(pq, (0, gate))
        dist[gate] = 0
    while pq:
        cost, u = heapq.heappop(pq)
        if cost > dist[u]:
            continue
        if u in summits:    # summit 도달하면 탐색 중단
            continue
        for v, w in hiking[u]:
            new_cost = max(cost, w)
            if new_cost < dist[v]: 
                dist[v] = new_cost
                heapq.heappush(pq, (new_cost, v))
                
    summits.sort()  # 번호 작은 것부터 확인
    
    result_summit = 0
    result_intensity = float('inf')
    
    for summit in summits:
        if dist[summit] < result_intensity:
            result_intensity = dist[summit]
            result_summit = summit
    
    return [result_summit, result_intensity]
"""
import heapq

def dijkstra(start, graph, n):
    dist = [float('inf')] * (n + 1)
    dist[start] = 0

    pq = [(0, start)] # 거리, 노드

    while pq:
        cost, u = heapq.heappop(pq)
        if cost > dist[u]:
            continue

        for v, w in graph[u]:
            new_cost = cost + w
            if new_cost < dist[v]:
                dist[v] = new_cost
                heapq.heappush(pq, (new_cost, v))

    return dist
"""

File: Hyunwoo/src/등산코스정하기.java
import java.util.*;

class Solution {

    boolean[] canGo;
    List<List<Node>> adj = new ArrayList<>();
    boolean[] isGate;
    boolean[] isSummit;
    int[] maxWeight;

    static class Node {
        int to;
        int weight;
        public Node(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }

    public int[] solution(int n, int[][] paths, int[] gates, int[] summits) {
        maxWeight = new int[n+1];
        for (int i = 0; i <= n; i++) {
            maxWeight[i] = Integer.MAX_VALUE;
        }

        isGate = new boolean[n+1];
        isSummit = new boolean[n+1];

        for (int gate : gates) {
            isGate[gate] = true;
        }

        for (int summit : summits) {
            isSummit[summit] = true;
        }

        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] path : paths) {
            int s = path[0];
            int e = path[1];
            int w = path[2];
            adj.get(s).add(new Node(e, w));
            adj.get(e).add(new Node(s, w));
        }

        return dijkstra(gates, summits);
    }

    public int[] dijkstra(int[] gates, int[] summits) {
        Queue<Node> pq = new PriorityQueue<>((o1, o2) -> {
            return o1.weight - o2.weight;
        });

        for (int gate : gates) {
            maxWeight[gate] = 0;
            pq.offer(new Node(gate, 0));
        }

        while (!pq.isEmpty()) {
            Node currNode = pq.poll();
            int curr = currNode.to;

            if (isSummit[curr])
                continue;

            if (currNode.weight > maxWeight[curr])
                continue;

            for (Node nextNode : adj.get(curr)) {
                if (isGate[nextNode.to])
                    continue;

                int newWeight = Math.max(maxWeight[curr], nextNode.weight);
                if (newWeight < maxWeight[nextNode.to]) {
                    maxWeight[nextNode.to] = newWeight;
                    pq.offer(new Node(nextNode.to, newWeight));
                }
            }
        }

        Arrays.sort(summits);
        int tempSummit = -1;
        int tempIntensity = 10000001;
        for (int summit : summits) {
            if (maxWeight[summit] < tempIntensity) {
                tempSummit = summit;
                tempIntensity = maxWeight[summit];
            }
        }

        return new int[] {tempSummit, tempIntensity};

    }
}